components/org.wso2.carbon.identity.tag.manager.core/src/main/java/org/wso2/carbon/identity/tag/manager/core/internal/impl/TagManagementServiceImpl.java

```diff
@@ -0,0 +1,271 @@
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getMessage());
+        }
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        tag = tagManagementDAO.loadTag(tagUuid, tenantUuid);
+        if (tag == null) {
+            throw new TagServiceClientException(
+                    ErrorMessage.ERROR_CODE_TAG_DATA_DOES_NOT_EXIST.getCode(),
+                    String.format(ErrorMessage.ERROR_CODE_TAG_DATA_DOES_NOT_EXIST.getMessage(), tagUuid));
+        }
+        return tag;
+    }
+
+    @Override
+    public List<String> getTagTypes(int limit, int offset, String sortBy, String sortOrder) throws TagServiceException {
+
+        List<String> tagList;
+        handleSorting(sortBy,sortOrder);
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        tagList = tagManagementDAO.loadTagTypes(limit, offset, sortBy, sortOrder);
+        return tagList;
+    }
+
+    @Override
+    public TagListResult filterTags(int limit, int offset, String sortBy, String sortOrder, String filter)
+            throws TagServiceException {
+
+        TagListResult tagListResult;
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        if (limit < 0) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_LIMIT_VALUE.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_LIMIT_VALUE.getMessage());
+        }
+        if (offset < 0) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_OFFSET_VALUE.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_OFFSET_VALUE.getMessage());
+        }
+        handleSorting(sortBy,sortOrder);
+        //TODO do filter validation
+        tagListResult = tagManagementDAO.filterTags(limit, offset, null, null, filter, tenantUuid);
+        return tagListResult;
+    }
+
+    @Override
+    public void updateTag(String tagUuid, String name, String description, boolean isPubliclyVisible)
+            throws TagServiceException {
+
+        if (StringUtils.isBlank(tagUuid)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getMessage());
+        }
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        if (!tagManagementDAO.isExistingTagId(tagUuid)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_TAG_UUID.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_TAG_UUID.getMessage());
+        }
+        if (StringUtils.isBlank(name)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_NAME.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_NAME.getMessage());
+        }
```

Comment:
Don't we need to check the description is null?

Reply:
As description is not a mandatory field, it can be null.